trace_decoder/src/wire.rs

```diff
@@ -0,0 +1,314 @@
+        one_of((0u8, 1u8)), // header
+        repeat_till(1.., instruction, eof).map(|(it, _)| {
+            NonEmpty::<Vec<_>>::new(it).expect("repeat_till should ensure non-empty collection")
+        }),
+    )
+    .parse(input)
+    {
+        Ok(it) => Ok(it),
+        Err(e) => bail!("parse error at offset {}: {}", e.offset(), e.inner()),
+    }
+}
+
+/// Names are taken from the spec.
+/// Spec also requires sequences to be non-empty
+///
+/// CBOR supports unsigned integers up to `2^64 -1`[^1], so we use [`u64`]
+/// for directly read integers.
+///
+/// [^1]: <https://en.wikipedia.org/wiki/CBOR#Integers_(types_0_and_1)>
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Instruction {
+    Leaf {
+        key: NonEmpty<Vec<U4>>,
+        value: NonEmpty<Vec<u8>>,
+    },
+    // BUG: the spec has `EXTENSION` and `ACCOUNT_LEAF` inconsistent
+    // between the instruction list and encoding list
+    Extension {
+        key: NonEmpty<Vec<U4>>,
+    },
+    Branch {
+        mask: u64,
+    },
+    Hash {
+        raw_hash: [u8; 32],
+    },
+    Code {
+        raw_code: NonEmpty<Vec<u8>>,
+    },
+    AccountLeaf {
+        key: NonEmpty<Vec<U4>>,
+        nonce: Option<u64>,
+        // BUG: see decode site
+        balance: Option<U256>,
+        has_code: bool,
+        has_storage: bool,
+    },
+    SmtLeaf(SmtLeaf),
+    // BUG: see parse site
+    EmptyRoot,
+    NewTrie,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SmtLeaf {
+    pub node_type: SmtLeafType,
+    pub address: NonEmpty<Vec<u8>>,
+    pub value: NonEmpty<Vec<u8>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum SmtLeafType {
+    Balance,
+    Nonce,
+    Code,
+    Storage(NonEmpty<Vec<u8>>),
+    CodeLength,
+}
+
+/// A single place to swap out the error type if required
+type PResult<T> = winnow::PResult<T, winnow::error::ContextError>;
+
+fn instruction(input: &mut &[u8]) -> PResult<Instruction> {
+    let start = input.checkpoint();
+    let opcode = any(input)?;
+    // I don't like the `winnow::combinator::dispatch!` macro
+    match opcode {
+        0x00 => trace(
+            "leaf",
+            (key, cbor).map(|(key, value)| Instruction::Leaf { key, value }),
+        )
+        .parse_next(input),
+        0x01 => trace("extension", key.map(|key| Instruction::Extension { key })).parse_next(input),
+        0x02 => trace("branch", cbor.map(|mask| Instruction::Branch { mask })).parse_next(input),
+        0x03 => {
+            trace("hash", array.map(|raw_hash| Instruction::Hash { raw_hash })).parse_next(input)
+        }
+        0x04 => {
+            trace("code", cbor.map(|raw_code| Instruction::Code { raw_code })).parse_next(input)
+        }
+        0x05 => trace("account_leaf", account_leaf).parse_next(input),
+        // BUG: this opcode is is undocumented, but the previous version of
+        //      this code had it, and our tests fail without it
```

Comment:
```suggestion
        // BUG: this opcode is undocumented, but the previous version of
        //      this code had it, and our tests fail without it.
```

Reply:
Also may be worth opening a tracking issue for it.

Reply:
#349